scripts/x-api.js

```diff
@@ -0,0 +1,596 @@
+      var arr;
+      // Need to not push in segments that happen from multiple triggers during scrubbing
+      if (endTime !== playedSegmentsSegmentStart && Math.abs(endTime - playedSegmentsSegmentStart) > 1 ) {
+        // Don't run if called too closely to each other.
+        arr = playedSegments == "" ? [] : playedSegments.split("[,]");
+        arr.push(formatFloat(playedSegmentsSegmentStart) + "[.]" + formatFloat(endTime));
+        playedSegments = arr.join("[,]");
+        playedSegmentsSegmentEnd = endTime;
+        playedSegmentsSegmentStart = null;
+      }
+    };
+
+    /**
+     * Append extra data to the XAPI statement's default object definition.
+     *
+     * @private
+     * @returns {Object} "Object" portion of JSON xAPI statement
+     */
+    var getXAPIObject = function () {
+      var event = new H5P.XAPIEvent();
+      event.setObject(videoInstance);
+      var xAPIObject = event.data.statement.object;
+
+      // Add definition type (required by xAPI Video Profile).
+      // @see https://liveaspankaj.gitbooks.io/xapi-video-profile/content/statement_data_model.html#241-definition
+      xAPIObject.definition.type = "https://w3id.org/xapi/video/activity-type/video";
+
+      // Add definition description (if video has a description).
+      if (videoInstance.contentId && H5PIntegration && H5PIntegration.contents && H5PIntegration.contents['cid-' + videoInstance.contentId].jsonContent) {
+        var videoData = JSON.parse(H5PIntegration.contents['cid-' + videoInstance.contentId].jsonContent);
+        if (videoData && videoData.interactiveVideo && videoData.interactiveVideo.video && videoData.interactiveVideo.video.startScreenOptions && videoData.interactiveVideo.video.startScreenOptions.shortStartDescription) {
+          xAPIObject.definition.description = {
+            "en-US": videoData.interactiveVideo.video.startScreenOptions.shortStartDescription
+          };
+        }
+      }
+
+      return xAPIObject;
+    };
+  }
+
+  /**
+  * Generate a random GUID string used for seesionID with video xAPI statements.
+  *
+  * @private
+  */
+  var guid = function () {
+    var s4 = function () {
+      return Math.floor((1 + Math.random()) * 0x10000).toString(16).substring(1);
+    };
+
+    return s4() + s4() + '-' + s4() + '-' + s4() + '-' + s4() + '-' + s4() + s4() + s4();
+  };
+
+  /**
+   * Format parameter as float (or null if invalid).
+   * Used when making arguments sent with video xAPI statments.
+   *
+   * @private
+   * @param {string} Number to convert to float
+   */
+  var formatFloat = function (number) {
+    if (number == null) {
+      return null;
+    }
+
+    return +(parseFloat(number).toFixed(2));
+  };
+
+  /**
+   * Convert duration in seconds to an ISO8601 duration string.
+   *
+   * @private
+   * @param {Number} time Duration in seconds
+   * @returns {String} Duration in ISO8601 duration format
+   */
+  var secondsToISO8601Duration = function (time) {
```

Comment:
`secondsTiISO8601Duration` does not work as intended. It doesn't return a string defined by the ISO8601 standard. The unit specific segments are returned as absolute totals, not as the "remainders".

For example, the input value of 8192 (seconds) should return PT2H16M32S for 2 hours, 16 minutes and 32 seconds, but it returns PT2H136M8192S.

Also, there are some missing semicolons, time is already defined, and I am not sure why `hasOwnProperty()`` is necessary if you're looping over the properties that you just defined yourself?

Reply:
I can't reproduce your comment on `secondsToISO8601Duration()`; calling `secondsToISO8601Duration(8192)` returns a value of `"PT2H16M32S"`, not sure where your value came from.

Agreed on `hasOwnProperty()`.